### ISSUE TITLE
Proposal: Compiler flag option to prevent assignment of any to a specifically typed prop

### ISSUE THREAD

**Issue by bakesteve** [microsoft/TypeScript#13539]
**TypeScript Version:**  2.1.1 / nightly (2.2.0-dev.201xxxxx)

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
let test = 'Hello';
let sneak: any = 4;
let cheater = 4;
test = cheater; // not allowed
test = sneak; // OK
```

**Expected behavior:**
with a compiler option (that defaults to false), it would be good to prevent this
If enabled, this flag, would throw the same error for the assign from `sneak` as it does for `cheater`

**Actual behavior:**
It allows the assignment from the `any` type (as per the language spec), but defeating our type safety

I know that all types are sub types of `any` so not sure how feasible it would be, but in cases where I cant easily avoid a type being set to any (`action.payload` coming from a middleware in my specific case. at least till we move to typed actions) it would be very helpful to be allowed to prevent this. And would still mean I could  cast from `any` if I wanted to then move from a `nay` to a strongly typed object

**Comment by mhegazy** [comment]
This is what `any` means really. may be what you are looking for is something like https://github.com/Microsoft/TypeScript/issues/13316 (and subsequently https://github.com/Microsoft/TypeScript/issues/13473) to allow you to remove all explicit `any` declarations?

**Comment by Namek** [comment]
the me, this is a worse problem:

```ts
let a: any = "text"
let b: number

b = a // that's OK but I'd like not to be without explicit cast
```

**Comment by mhegazy** [comment]
In your editor, find all references on `any`, this should list all the ones you have in your project.

**Comment by Namek** [comment]
Well, no. I want the compiler to throw errors at me. This shouldn't be even possible without explicit cast. I do have some use cases of `any` that I can't change.

**Comment by mhegazy** [comment]
> Well, no. I want the compiler to throw errors at me. This shouldn't be even possible without explicit cast. I do have some use cases of any that I can't change.

why not change these cases to `object`? this will force you to cast when use them.

**Comment by Namek** [comment]
Thanks. I'll do that after upgrade.

**Comment by mhegazy** [comment]
> However, that won't help with generic types probably.

generic type inference never produces `any` if it was not already there. if inference found no candidates it defaults to the upper bound (i.e. `{}`). 

So the only two sources of `any` in your compilation are 1. implicit any for un-annotated declarations (should be all flagged with `--noImplicitAny` as errors), and 2. explicit `any` annotations (in your code, or references .d.ts files), these should be detectable by find-all-references in your IDE with TS 2.2 and later.

**Comment by RyanCavanaugh** [comment]
Discussed at #24737; we don't have a good way forward on this.

**Comment by stefee** [comment]
For anyone finding this now and wondering if there is a solution, you can use typescript-eslint with the following rules:

```json
    "@typescript-eslint/no-unsafe-argument": "error",
    "@typescript-eslint/no-unsafe-assignment": "error",
    "@typescript-eslint/no-unsafe-call": "error",
    "@typescript-eslint/no-unsafe-member-access": "error",
    "@typescript-eslint/no-unsafe-return": "error"
```

Links to the documentation for these rules can be found here: https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#supported-rules

You must also add this to your eslint config to allow the eslint plugin find your ts project configuration:

```json
  "parserOptions": {
    "project": "./tsconfig.json"
  },
```
